DESIGN.md

```diff
@@ -0,0 +1,400 @@
+type ByteSigner
+type ByteVerifier
+type Headers
+    func (h *Headers) Decode(o []interface{}) (err error)
+    func (h *Headers) DecodeProtected(o interface{}) (err error)
+    func (h *Headers) DecodeUnprotected(o interface{}) (err error)
+    func (h *Headers) EncodeProtected() (bstr []byte)
+    func (h *Headers) EncodeUnprotected() (encoded map[interface{}]interface{})
+type KeyType
+type RSAOptions
+type Sign1Message
+    func NewSign1Message() *Sign1Message
+    func (message *Sign1Message) MarshalCBOR() ([]byte, error)
+    func (m *Sign1Message) SigStructure(external []byte) ([]byte, error)
+    func (m *Sign1Message) Sign(rand io.Reader, external []byte, signer Signer) (err error)
+    func (message *Sign1Message) UnmarshalCBOR(data []byte) (err error)
+    func (m Sign1Message) Verify(external []byte, verifier Verifier) (err error)
+type SignMessage
+    func NewSignMessage() *SignMessage
+    func (m *SignMessage) AddSignature(s *Signature)
+    func (message *SignMessage) MarshalCBOR() ([]byte, error)
+    func (m *SignMessage) SigStructure(external []byte, signature *Signature) (ToBeSigned []byte, err error)
+    func (m *SignMessage) Sign(rand io.Reader, external []byte, signers []Signer) (err error)
+    func (message *SignMessage) UnmarshalCBOR(data []byte) (err error)
+    func (m *SignMessage) Verify(external []byte, verifiers []Verifier) (err error)
+type Signature
+    func NewSignature() (s *Signature)
+    func (s *Signature) Decode(o interface{})
+    func (s *Signature) Equal(other *Signature) bool
+type Signer
+    func NewSigner(alg *Algorithm, options interface{}) (signer *Signer, err error)
+    func NewSignerFromKey(alg *Algorithm, privateKey crypto.PrivateKey) (signer *Signer, err error)
+    func (s Signer) GetAlg() *Algorithm
+    func (s *Signer) Public() (publicKey crypto.PublicKey)
+    func (s *Signer) Sign(rand io.Reader, digest []byte) (signature []byte, err error)
+    func (s *Signer) Verifier() (verifier *Verifier)
+type Verifier
+    func (v *Verifier) Verify(digest []byte, signature []byte) (err error)
+```
+
+Here are the proposed APIs:
+
+```go
+const (
+    HeaderLabelAlgorithm         = 1
+    HeaderLabelCritical          = 2
+    HeaderLabelContentType       = 3
+    HeaderLabelKeyID             = 4
+    HeaderLabelCounterSignature  = 7
+    HeaderLabelCounterSignature0 = 9
+    HeaderLabelX5Bag             = 32
+    HeaderLabelX5Chain           = 33
+    HeaderLabelX5T               = 34
+    HeaderLabelX5U               = 35
+)
+
+const (
+    CBORTagSignMessage  = 98
+    CBORTagSign1Message = 18
+)
+
+const (
+    ContextSignature  = "Signature"
+    ContextSignature1 = "Signature1"
+)
+
+var (
+    AlgorithmPS256   *Algorithm
+    AlgorithmPS384   *Algorithm
+    AlgorithmPS512   *Algorithm
+    AlgorithmES256   *Algorithm
+    AlgorithmES384   *Algorithm
+    AlgorithmES512   *Algorithm
+    AlgorithmEd25519 *Algorithm
+)
+
+type Algorithm struct {

```

Comment:
I'm seeing `Algorithm` passed by reference in several APIs. I think it would be better to always pass it by value. Reasoning:
- In a 64-bit system, `Algorithm` size is only 32 bytes, so moving it around just copying it is cheap and fast, and it avoids allocations!
- AFAIU, an Algorithm is not intended to be mutated, it must have well-known predefined properties which are set once and then passed around.
- Passing it by value reduces the chances that someone modifies its properties and breaks unrelated code which reuse the same pointer.